### PR TITLE
Global request timeouts #989

### PR DESCRIPTION
**Pull request by hvltaj**
Adds timeout to global_request in Transport.
Tests included.

Implements #989

### CODE REVIEW

File: tests/test_transport.py [hunk 1]
```diff
@@ -927,3 +929,49 @@ def test_channel_send_memoryview(self):
...
             self.assertEqual(sfile.read(len(data)), data)
+
+    def test_global_request_timeout(self):
+        """
+        verify global_request timeouts.
+        """
+        self.setup_test_server()
+
+        with mock.patch.object(self.server, 'check_port_forward_request',
+                               new=lambda x, y: time.sleep(2)):
```
Comment (bz2) [inline tests/test_transport.py]: Should not use `time.sleep()` in tests. With the real timeout of 1 second below, that ensures this test will always be slow. Can either mock `time.time()` in the function for the same effect, or pass in a timer at some point.

File: paramiko/transport.py [hunk 2]
```diff
@@ -1021,10 +1021,16 @@ def global_request(self, kind, data=None, wait=True):
         :param bool wait:
             ``True`` if this method should not return until a response is
             received; ``False`` otherwise.
+        :param float timeout:
+            optional timeout making a global request, no default set
```
Comment (bz2) [inline paramiko/transport.py]: I realise this is somewhat modelled on `open_channel()`, but having both `wait=` and `timeout=` is weird. Timeout is a superset of the wait behaviour, but means nothing without it.

**Comment by philusky** [comment]
Timeout can be passed in the ''wait'' argument now.
I set up the default timeout to 5s, it will work as well if someone passes 'wait=True'.
I mocked time.time() calls to be 6 second apart.

**Comment by ploxiln** [comment]
You should rebase/squash such that only a couple relevant git commits appear in this PR.

File: paramiko/transport.py [hunk 3]
```diff
@@ -1038,13 +1043,17 @@ def global_request(self, kind, data=None, wait=True):
         self._send_user_message(m)
         if not wait:
             return None
-        while True:
+        # sets default timeout to 5 in case ``wait`` == True
+        timeout = 5 if isinstance(wait, bool) else wait
```
Comment (ploxiln) [inline paramiko/transport.py]: This is a bit awkward - there's no elegant value to mean "forever" (which might be useless), and 5 is written in two places.

Here's another idea: let `True` mean "forever", and change all the places in paramiko.transport which call `global_request(... wait=True)` to just not specify wait, allowing the default to be used. Then, perhaps in a later PR, the appropriate timeout configured in the Transport could be specified when `global_request()` is called.

Comment (ploxiln) [reply paramiko/transport.py]: Also, the other timeouts in Transport have higher default values, 15 or 30 seconds.

**Comment by bskinn** [comment]
A number of timeout related features were added in paramiko 3.1.

This is sufficiently old that I'm going to close it. Please check paramiko 3.1 or later—if the timeout you want still isn't implemented, please open a new issue/PR and reference this one.

Thanks!

**Pull request closed without merging**